### ISSUE TITLE
Custom template names for string files

### ISSUE THREAD

**Issue by potmo** [AliSoftware/SwiftGen#242]
It would be nice to be able to use custom template names.

Swiftgen currently only support the standard c-style unnamed `%d` templates etc.
In the project I'm currently working on we use `${someName}` instead (and manually create the replacement code etc that swifgen does).

It would be really cool to be able to use named arguments like
`This is some string file translation for ${someName} with ${somethingElse}`

would become:
`let wow = tr(.TheString(someName: "a file", somethingElse: "the named stuff"))`

Is that something that anyone would appreciate? In that case I could take a stab at it (since I really need it).

**Comment by djbe** [comment]
It'd definitely be welcome!

Do take into account you'll probably have to modify the data structure of the parser, and thus the context for template rendering. If at all possible, it'd be best if you could do this in a non-breaking way for existing templates.

**Comment by AliSoftware** [comment]
Interesting idea :+1: but how would you then specify the type for that custom param name?

Like if I want the calling code to look like `tr(.theString(someName: "Paul", count: 5))`, how would the string in `Localizable.strings` look like to specify that parameter `count` expects an `Int`, not a `String`?

**Comment by djbe** [comment]
Ah, I didn't notice these are custom string templates, not some built in strings functionality we're missing. Don't know how you're going to combine this with existing "standard" functionality.

**Comment by djbe** [comment]
I'd try to work with some combination of:
"your.custom.key.format.here(param1, param2)" =

**Comment by potmo** [comment]
I think that it's reasonable to scope down `${name}` to only be strings and not support other formats.

Im thinking I could add a new type to `PlaceholderType` (https://github.com/AliSoftware/SwiftGen/blob/master/GenumKit/Parsers/StringsFileParser.swift#L53). That new format would have to contain an associated enum value `name` witch would in turn make the enum not use `String` as `rawValue`. 

From what I understand the `names` are already exposed as template variables so templates will not have to change and can be backwards compatible. 

The problem will arise when the templates needs to `translatedString.replacingOccurrences(of: "${customKey}", with: dynamicVariable)`

The ordering of the arguments needs to be deterministic so maybe I need to order them alphabetically. 

I think I might try and realise that it's hard to implement this without breaking everything else.
Is it an option to add a flag to the cli command like `--custom-template-pattern ${(.*?)}` or `--template-pattern c-style|named` to enable this feature?

**Comment by AliSoftware** [comment]
I honestly am not a fan of adding a custom CLI flag that specific and only usable for a specific template. I'm even considering reverting the `--enumName` flag in future versions.

In exchange, I considered adding a more generic flag instead, allowing you to pass arbitrary values to the Stencil context, like `--param foo=bar --param enumName=Loc --param stringPattern="${(.*?)}` etc. I think we discussed that with @djbe (or was it with @peterlivesey for #188 ?) but never took the time to actually implement that. Might be the good occasion ?

**Comment by potmo** [comment]
Yes that would at least give me the opportunity to write exotic templates that would support my usecase. I'll see if I get the time to give the named pattern a go without breaking any backwards compatibility.

**Comment by djbe** [comment]
@potmo the parameter parsing is available since SwiftGen 4.2, so you could get started with that. If you can find the time to implement this, it'll certainly be welcome as long as it won't break other templates and current parsing behaviours.

If you'd like an easier starter task, related to this command, can I recommend #39 (string tablename support). There even is a PR related to it on which you can base yourself (#86), although it is quite outdated.

**Comment by Dschee** [comment]
I'd love to see and help with this feature as well, I just opened a duplicate issue where I laid out my expected usage, I'd like to quote it here for keep all the content within this issue:

> I was just using a key like `MAX_SPEED_EXCEEDED_NOTIFICATION.MESSAGE` in my code which generates to `L10n.MaxSpeedExceededNotification.message(50, "mi/h")` on the usage side. The value in the Strings file is something like `Max speed of %d %@ exceeded`.
> 
> It is not clear though for both the translator and also the programmer what these placeholders stand for. Both need to guess some part of it.
> 
> Natural Approach
> ---
> Wouldn't it be better if we could write something like:
> 
> ```
> MAX_SPEED_EXCEEDED_NOTIFICATION.MESSAGE(MAX_SPEED_VALUE,MAX_SPEED_UNIT)
> ```
> 
> The translator would immediately know what the two stand for and could look up the correct order in the source translation (e.g. English). We could build the given parentheses structure into a template so the usage in code would look like this: 
> ```
> L10n.MaxSpeedExceededNotification.message(maxSpeedValue: 50, maxSpeedUnit: "mi/h")
> ```
> This would also make sure every programmer using the API or reading the code knows exactly what's going on.
> 
> Extensive Approach
> ---
> Alternatively we could also build our own replacement logic. For example, using the common environment variable prefix `$` a key could look something like this: 
> ```
> MAX_SPEED_EXCEEDED_NOTIFICATION.MESSAGE($MAX_SPEED_VALUE,$MAX_SPEED_UNIT)
> ```
> 
> Then the translation value could even be easier for a translator to manage:
> ```
> Max speed of $MAX_SPEED_VALUE $MAX_SPEED_UNIT exceeded
> ```
> The translations would also be much more readable. Of course we would not only write a new template, but also a replacement logic into the SwiftGenKit library that would automatically search a translation for the placeholders defined in the key and replace them accordingly. Usage would then be the same as in the above idea:
> ```
> L10n.MaxSpeedExceededNotification.message(maxSpeedValue: 50, maxSpeedUnit: "mi/h")
> ```
> The main advantage of this alternative approach would be that translators wouldn't have to deal with things like `%@` or `%d`.
